mixer/test/e2e/e2e_reftracking_test.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"strconv"
+	"strings"
+	"testing"
+
+	"google.golang.org/grpc"
+
+	istio_mixer_v1 "istio.io/api/mixer/v1"
+	"istio.io/istio/mixer/pkg/config/storetest"
+	testEnv "istio.io/istio/mixer/pkg/server"
+	spyAdapter "istio.io/istio/mixer/test/spyAdapter"
+	e2eTmpl "istio.io/istio/mixer/test/spyAdapter/template"
+)
+
+const (
+	refTrackingGlobalCfg = `
+apiVersion: "config.istio.io/v1alpha2"
+kind: attributemanifest
+metadata:
+  name: istio-proxy
+  namespace: default
+spec:
+    attributes:
+      destination.service:
+        value_type: STRING
+      source.service:
+        value_type: STRING
+      request.headers:
+        value_type: STRING_MAP
+---
+`
+	refTrackingSvcCfg = `
+apiVersion: "config.istio.io/v1alpha2"
+kind: fakeHandler
+metadata:
+  name: fakeHandlerConfig
+  namespace: istio-system
+
+---
+
+apiVersion: "config.istio.io/v1alpha2"
+kind: samplecheck
+metadata:
+  name: checkInstance
+  namespace: istio-system
+spec:
+  stringPrimitive: "destination.service"
+---
+
+apiVersion: "config.istio.io/v1alpha2"
+kind: rule
+metadata:
+  name: rule1
+  namespace: istio-system
+spec:
+  match: destination.service == "echosrv1.istio.svc.cluster.local" && source.service == "foo"
+#  match: destination.service == "echosrv1.istio.svc.cluster.local" && request.headers["x-request-id"] == "foo"
```

Comment:
Should add (additional) test that uses `request.headers[]` the stuff you have commented out here.
It exercises a different code path.